### PR TITLE
Work around mypy issue #4554 in attr.Factory

### PR DESCRIPTION
**Pull request by euresti**
This makes the following not give a type error because of python/mypy#4554
```
   y: DefaultDict[str, List[int]] = Factory(lambda: defaultdict(list))
```

### CODE REVIEW

**Comment by euresti** [comment]
@chadrik Another change for you.  :)

File: third_party/2and3/attr/__init__.pyi [hunk 1]
```diff
@@ -21,7 +21,13 @@ _ValidatorArgType = Union[_ValidatorType[_T], Sequence[_ValidatorType[_T]]]
 NOTHING: object
 
 # NOTE: Factory lies about its return type to make this possible: `x: List[int] = Factory(list)`
-def Factory(factory: Union[Callable[[], _T], Callable[[Any], _T]], takes_self: bool = ...) -> _T: ...
+# Work around mypy issue #4554 by using an overload vs a Union.
+@overload
+def Factory(factory: Callable[[], _T]) -> _T: ...
```
Comment (JelleZijlstra) [inline third_party/2and3/attr/__init__.pyi]: Isn't this equivalent to the second overload? If not why, don't you need one where the Callable takes an argument and the second argument is omitted?

Comment (euresti) [reply third_party/2and3/attr/__init__.pyi]: If you don't set `takes_self` then you'd better not send in a function with an argument.  If you do set `takes_self` then both are allowed.

That being said I'm ok with switching this to:

```
@overload
def Factory(factory: Callable[[], _T], takes_self: bool = ...) -> _T: ...
@overload
def Factory(factory: Callable[[Any], _T], takes_self: bool = ...) -> _T: ...
```
And forget about the first overload.

Comment (JelleZijlstra) [reply third_party/2and3/attr/__init__.pyi]: Yes, the first overload is redundant because the argument has a default.

**Comment by euresti** [comment]
First overload removed.

**Pull request merged**